torch/nn/utils/no_init.py

```diff
@@ -0,0 +1,60 @@
+from typing import Type
+
+from ..._C import _suspend_backend_dispatch, _restore_backend_dispatch
+from ..modules import Module
+
+
+def no_init(module_cls: Type[Module], *args, **kwargs) -> Module:
+    """Constructs a module without initializing its parameters and buffers.
+
+    This function is meant to be used if the size of the specified module is big
+    and you want to inspect it without fully instantiating. For instance an
+    automated sharding system can construct a module with ``no_init()``, inspect
+    its architecture, and materialize its layers at a later time.
+
+    Internally ``no_init()`` forces all parameters and buffers of the module and
+    its sub-modules to use a meta device regardless of the requested device
+    type. The returned module can be used for inspection purposes, but has to be
+    materialized by calling its ``materialize()`` method before doing actual
+    work.
+
+    Args:
+        module_cls:
+            The type of the module to construct.
+        args:
+            The positional arguments to pass to the module's constructor.
+        kwargs:
+            The keyword arguments to pass to the module's constructor.
+
+    Returns:
+        A module instance with uninitialized parameters and buffers.
+
+    Example::
+        >>> import torch
+        >>> m = torch.nn.utils.no_init(torch.nn.Linear, 5, 1)
+        >>> m.weight
+        Parameter containing:
+        tensor(..., device='meta', requires_grad=True)
+        >>> n = m.materialize()
+        >>> n.weight
+        Parameter containing:
+        tensor([[-1.4677e+24,  4.5915e-41,  1.4013e-45,  0.0000e+00, -1.4677e+24,
+                  4.5915e-41]], requires_grad=True)
+
+    Note:
+        The ``args`` and ``kwargs`` arguments must be treated as immutable since
+        they will be used a second time during materialization.
+    """
+    _suspend_backend_dispatch()
```

Comment:
would I be correct if I assume `_suspend_backend_dispatch` and `_restore_backend_dispatch` can also be wrapped into a context manager? And users can do sth like:

```python
with torch.distributed.nn.no_init():
  module = MyModule(...)
``` 

And we can provide a separate `torch.distributed.nn.materialize()` API to materialize the model?

Reply:
Yes, technically there is nothing preventing us from exposing this future as a context manager. It is more an API design question though. Since `no_init()` is quite intrusive and changes the behavior of the library in a spurious way (the intent and the outcome do not match: the caller "initializes" the module in a regular way, but ends up with a meta module), I wanted to artificially limit the scope where it can be used. If it was up to me, I would even return an opaque object instead of a module from the `no_init()` call. That opaque object would publicly expose only a `materialize()` function that returns a newly-constructed `nn.Module`. All other related large model PyTorch APIs would use an internal interface of the opaque object for introspection.

Again, as I mentioned, this was purely a design choice that I made. If the consensus is to expose it as a context manager, I am also fine with it.

Reply:
> If it was up to me, I would even return an opaque object instead of a module from the no_init() call.

+1 to this. And we can move this API from `torch.nn` to `torch.distributed.nn`, which is only designed for distributed training + large models where the users know what they are doing.